presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationSubqueryToLeftOuterJoin.java

```diff
@@ -0,0 +1,412 @@
+import com.google.common.collect.ImmutableSet;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static com.facebook.presto.sql.planner.optimizations.PlanNodeUtils.findFirstPlanNodeMatching;
+import static com.facebook.presto.sql.planner.optimizations.PlanNodeUtils.removeFirstPlanNodesMatching;
+import static com.facebook.presto.sql.planner.optimizations.PlanNodeUtils.replaceAllPlanNodesMatching;
+import static com.facebook.presto.sql.planner.plan.SimplePlanRewriter.rewriteWith;
+import static com.facebook.presto.sql.tree.BooleanLiteral.TRUE_LITERAL;
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableMap;
+import static java.util.Objects.requireNonNull;
+
+public class TransformCorrelatedScalarAggregationSubqueryToLeftOuterJoin
+        implements PlanOptimizer
+{
+    private final Metadata metadata;
+
+    public TransformCorrelatedScalarAggregationSubqueryToLeftOuterJoin(Metadata metadata)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+    }
+
+    @Override
+    public PlanNode optimize(
+            PlanNode plan,
+            Session session,
+            Map<Symbol, Type> types,
+            SymbolAllocator symbolAllocator,
+            PlanNodeIdAllocator idAllocator)
+    {
+        return rewriteWith(new Rewriter(idAllocator, symbolAllocator, metadata), plan, null);
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<PlanNode>
+    {
+        private final PlanNodeIdAllocator idAllocator;
+        private final SymbolAllocator symbolAllocator;
+        private final Metadata metadata;
+        private final ScalarAggregationSourceRewriter scalarSourceRewriter;
+
+        public Rewriter(PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Metadata metadata)
+        {
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+            this.symbolAllocator = requireNonNull(symbolAllocator, "symbolAllocator is null");
+            this.metadata = requireNonNull(metadata, "metadata is null");
+            this.scalarSourceRewriter = new ScalarAggregationSourceRewriter(idAllocator);
+        }
+
+        @Override
+        public PlanNode visitApply(ApplyNode node, RewriteContext<PlanNode> context)
+        {
+            ApplyNode rewrittenNode = (ApplyNode) context.defaultRewrite(node, context.get());
+            if (!rewrittenNode.getCorrelation().isEmpty()) {
+                Optional<AggregationNode> scalarAggregation = findFirstPlanNodeMatching(
+                        rewrittenNode.getSubquery(),
+                        AggregationNode.class::isInstance,
+                        Predicates.<PlanNode>isInstance(ProjectNode.class).or(EnforceSingleRowNode.class::isInstance));
```

Comment:
I think this would be more readable as:

```java
e -> e instanceof ProjectNode || e instanceof EnforceSingleRowNode
```

or introduce a helper method to construct a predicate like:

```java
Predicates.instanceOfAny(ProjectNode.class, EnfoceSingleRowNode.class)
```